### PR TITLE
Enable offline migration mode on sql migration extension

### PR DESCRIPTION
**Pull request by rchlkm**
This PR adds the offline migration mode functionality to the SQL Migration extension.

**1. Migration Mode Page**
User can select offline migration
image

**2. Database Backup Page**
For Offline Blob Migrations, user will need to select a last backup file (new column added).
Updated table so that each dropdown only fetches data if the previous column has already been selected. Dropdowns are also disabled if previous column hasn't been selected.
Add placeholder text (i.e. Select a blob container value first) in disable dropdown with tooltip.
image

image

**3. Summary Page**
Add Last Backup File column for online blob migrations
image

**4. Migration Cutover Page**
Hide Complete Cutover button for offline migrations since offline migrations are "auto cutover".
Hide Backup File table for offline migrations (and online blob migrations)
Hide Full Backup, Last LSN, Last Backup File Timestamp items from the grid for offline migrations
Add status icon next to the Migration Status

Offline successful example:
image

Online failed example:
image

### CODE REVIEW

**Review by aasimkhan30 — APPROVED** [review]

File: src/sql/workbench/browser/modelComponents/dropdown.component.ts [hunk 1]
```diff
@@ -145,6 +145,10 @@ export default class DropDownComponent extends ComponentBase<azdata.DropDownProp
 			}
 			this._editableDropdown.enabled = this.enabled;
 			this._editableDropdown.fireOnTextChange = this.fireOnTextChange;
+
+			if (this.showInputTooltip) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/browser/modelComponents/dropdown.component.ts:149]: Please submit this is a separate PR. Changes to core components (especially ones as widely used as dropdowns) should be reviewed separately to ensure they're properly reviewed.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: src/sql/workbench/browser/modelComponents/dropdown.component.ts (hunk 1 above)
Comment (rchlkm) [reply src/sql/workbench/browser/modelComponents/dropdown.component.ts:149]: I've created a separate PR for the dropdown component. Once the other PR is merged, I will update this one.
https://github.com/microsoft/azuredatastudio/pull/16468

File: extensions/sql-migration/src/constants/strings.ts [hunk 2]
```diff
@@ -129,11 +129,23 @@ export const NO_LOCATION_FOUND = localize('sql.migration.no.location.found', "No
 export const NO_STORAGE_ACCOUNT_FOUND = localize('sql.migration.no.storageAccount.found', "No storage account found");
 export const NO_FILESHARES_FOUND = localize('sql.migration.no.fileShares.found', "No file shares found");
 export const NO_BLOBCONTAINERS_FOUND = localize('sql.migration.no.blobContainers.found', "No blob containers found");
+export const NO_FILE_NAMES_FOUND = localize('sql.migration.no.fileNames.found', "No file names found");
```
Comment (brian-harris) [inline extensions/sql-migration/src/constants/strings.ts:132]: should this be "No files found"

File: extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialog.ts [hunk 3]
```diff
@@ -375,9 +384,12 @@ export class MigrationCutoverDialog {
 			await this.refreshStatus();
 		}));
 
-		headerActions.addItem(this._cutoverButton, {
-			flex: '0'
-		});
+		if (this._isProvisioned() && this._isOnlineMigration()) {
+			// _todo: should we just not create the cutover button if offline?
+			headerActions.addItem(this._cutoverButton, {
```
Comment (brian-harris) [inline extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialog.ts:389]: I agree it would be good to only create the controls that are needed

File: extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialog.ts [hunk 4]
```diff
@@ -536,19 +548,22 @@ export class MigrationCutoverDialog {
...
-						lastLSN: activeBackupSet.lastLSN
+				if (this._shouldDisplayBackupFileTable()) {
+					tableData.push(
+						{
+							fileName: activeBackupSet.listOfBackupFiles[0].fileName,
+							type: activeBackupSet.backupType,
+							status: activeBackupSet.listOfBackupFiles[0].status,
+							dataUploaded: `${convertByteSizeToReadableUnit(activeBackupSet.listOfBackupFiles[0].dataWritten)}/ ${convertByteSizeToReadableUnit(activeBackupSet.listOfBackupFiles[0].totalSize)}`,
+							copyThroughput: (activeBackupSet.listOfBackupFiles[0].copyThroughput / 1024).toFixed(2),
+							backupStartTime: activeBackupSet.backupStartDate,
```
Comment (brian-harris) [inline extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialog.ts:559]: is the default for copyThroughput 0 or can it be undefined and cause .toFixed to throw an error

File: extensions/sql-migration/src/wizard/wizardController.ts [hunk 5]
```diff
@@ -79,7 +79,9 @@ export class WizardController {
 		});
 
 		await Promise.all(wizardSetupPromises);
-		await pages[0].onPageEnter();
+		wizard.onPageChanged(async (pageChangeInfo: azdata.window.WizardPageChangeInfo) => {
+			await pages[0].onPageEnter(pageChangeInfo);
```
Comment (brian-harris) [inline extensions/sql-migration/src/wizard/wizardController.ts:83]: did this get merged with latest, I was expecting to see this:
this.extensionContext.subscriptions.push(wizard.onPageChanged(async (pageChangeInfo: azdata.window.WizardPageChangeInfo) => {

File: extensions/sql-migration/src/wizard/databaseBackupPage.ts [hunk 6]
```diff
@@ -640,9 +652,20 @@ export class DatabaseBackupPage extends MigrationWizardPage {
...
-
+	public async onPageEnter(pageChangeInfo: azdata.window.WizardPageChangeInfo): Promise<void> {
 		if (this.migrationStateModel.refreshDatabaseBackupPage) {
+			const isOfflineMigration = this.migrationStateModel._databaseBackup?.migrationMode === MigrationMode.OFFLINE;
+			const lastBackupFileColumnIndex = this._blobContainerTargetDatabaseNamesTable.columns.length - 1;
+			this._blobContainerTargetDatabaseNamesTable.columns[lastBackupFileColumnIndex].hidden = !isOfflineMigration;
+
+			this._networkShareButton.checked = false;
+			this._networkTableContainer.display = 'none';
+			this._networkShareContainer.updateCssStyles({ 'display': 'none' });
```
Comment (brian-harris) [inline extensions/sql-migration/src/wizard/databaseBackupPage.ts:663]: is setting .display = 'none'  preferred over setting css styles to hide.

File: extensions/sql-migration/src/wizard/databaseBackupPage.ts [hunk 7]
```diff
@@ -799,13 +864,17 @@ export class DatabaseBackupPage extends MigrationWizardPage {
 				targetRow.push({
 					value: this._blobContainerDropdowns[index]
 				});
+				targetRow.push({
+					value: this._blobContainerLastBackupFileDropdowns[index]
+				});
 				data.push(targetRow);
 			});
 			this._blobContainerTargetDatabaseNamesTable.dataValues = data;
```
Comment (brian-harris) [inline extensions/sql-migration/src/wizard/databaseBackupPage.ts:872]: To improve reliability, I suggest changing this to:
await this._blobContainerTargetDatabaseNamesTable.setDataValues(data);

File: extensions/sql-migration/src/constants/strings.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rchlkm) [reply extensions/sql-migration/src/constants/strings.ts]: updated to "No blob files found"

Comment (rchlkm) [reply extensions/sql-migration/src/constants/strings.ts]: updated to "No blob files found"

File: extensions/sql-migration/src/wizard/databaseBackupPage.ts (hunk 7 above)
Comment (rchlkm) [reply extensions/sql-migration/src/wizard/databaseBackupPage.ts]: updated 
from
`this._blobContainerTargetDatabaseNamesTable.dataValues = data;`
to
`await this._blobContainerTargetDatabaseNamesTable.setDataValues(data);`

File: extensions/sql-migration/src/wizard/databaseBackupPage.ts (hunk 6 above)
Comment (rchlkm) [reply extensions/sql-migration/src/wizard/databaseBackupPage.ts:668]: I tried various combinations of using only .display = 'none' and using updateCssStyles, but for some reason, this setup was the only one that produced the expected behavior.

File: extensions/sql-migration/src/wizard/wizardController.ts (hunk 5 above)
Comment (rchlkm) [reply extensions/sql-migration/src/wizard/wizardController.ts:86]: merged with the latest and resolved this issue

File: extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialog.ts (hunk 4 above)
Comment (rchlkm) [reply extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialog.ts:564]: I added a check to see if activeBackupSet.listOfBackupFiles[0].copyThroughput is null

File: extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialog.ts (hunk 3 above)
Comment (rchlkm) [reply extensions/sql-migration/src/dialog/migrationCutover/migrationCutoverDialog.ts]: I wanted to create the necessary controls only, but I realized that there are cases where we can't properly determine if it's necessary or not.

While the migration is being created (provisioningStatus=Creating), we can't determine if the migration is online or offline.
For the cutover button, I decided to hide it by default. One we can determine the migration mode, the button will show up or remain hidden.

**Review by aasimkhan30 — APPROVED** [review]

**Comment by aasimkhan30** [comment]
@rchlkm  Please remove the dropdown changes before merging the PR in

**Review by brian-harris — APPROVED** [review]

File: src/sql/workbench/browser/modelComponents/dropdown.component.ts (hunk 1 above)
Comment (rchlkm) [reply src/sql/workbench/browser/modelComponents/dropdown.component.ts]: removed the dropdown changes from this PR

**Review by Charles-Gagnon — APPROVED** [review]

File: extensions/sql-migration/src/constants/strings.ts [hunk 8]
```diff
@@ -270,6 +285,7 @@ export const SUMMARY_AZURE_STORAGE = localize('sql.migration.summary.azure.stora
 export const SUMMARY_IR_NODE = localize('sql.migration.ir.node', "Integration Runtime node");
 export const NETWORK_SHARE = localize('sql.migration.network.share', "Network Share");
 export const BLOB_CONTAINER = localize('sql.migration.blob.container.title', "Blob Container");
+export const BLOB_CONTAINER_LAST_BACKUP_FILE = localize('sql.migration.blob.container.last.backup.file.label', "Last Backup File");
```
Comment (alanrenmsft) [inline extensions/sql-migration/src/constants/strings.ts:288]: for label of controls, only the first letter of the first word needs be upper case. It would be good if you can clean up the places that are not following this convention.

**Review by alanrenmsft — APPROVED** [review]

File: extensions/sql-migration/src/constants/strings.ts (hunk 8 above)
Comment (rchlkm) [reply extensions/sql-migration/src/constants/strings.ts:288]: Next week, we will be revising the strings based on feedback from out content writer. We will address this then.
cc: @brian-harris
